jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/ContextHandlerFactory.java

```diff
@@ -0,0 +1,77 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.deploy;
+
+import java.nio.file.Path;
+import java.util.Set;
+
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.handler.ContextHandler;
+import org.eclipse.jetty.util.Attributes;
+import org.eclipse.jetty.util.component.Environment;
+
+/**
+ * A factory to create a ContextHandler from a set of associated files.
+ *
+ * @see DeploymentScanner
+ */
+public interface ContextHandlerFactory
+{
+    /**
+     * The attribute name for the value holding the specific class to used to create the ContextHandler
+     */
+    String CONTEXT_HANDLER_CLASS_ATTRIBUTE = "jetty.deploy.contextHandlerClass";
+    /**
+     * The attribute name for the value holding the default class to used for creating ContextHandlers in the environment
+     */
+    String CONTEXT_HANDLER_CLASS_DEFAULT_ATTRIBUTE = "jetty.deploy.default.contextHandlerClass";
+    /**
+     * The attribute name for the environment name.
+     */
+    String ENVIRONMENT_ATTRIBUTE = "environment";
+    /**
+     * Attributes that start with this name return a String pointing to XML on the filesystem that should be used for Environment based
+     * modifications to the ContextHandler that was created.
+     *
+     * <pre>
+     * jetty.deploy.environmentXml=config/base-ee8.xml
+     * jetty.deploy.environmentXml.auditing=config/auditing-ee8.xml
+     * </pre>
+     */
+    String ENVIRONMENT_XML_ATTRIBUTE = "jetty.deploy.environmentXml";
+    /**
+     * Attribute name that stores the {@code List<Path>} pointing to XML files that represent
+     * the Environment XML files to apply to the ContextHandler has been created.
+     */
+    String ENVIRONMENT_XML_PATHS_ATTRIBUTE = "jetty.deploy.paths.environmentXmls";
```

Comment:
Now that we have the `environments` directory, is this mechanism still necessary?

Reply:
The second one is almost an internal implementation detail, not to be used by users.